persistence-dse-6.8/src/main/java/io/stargate/db/dse/impl/DelegatingAuthorizer.java

```diff
@@ -0,0 +1,222 @@
+    return permissions;
+  }
+
+  @Override
+  public Set<Permission> revoke(
+      AuthenticatedUser performer,
+      Set<Permission> permissions,
+      IResource resource,
+      RoleResource revokee,
+      GrantMode... grantModes) {
+    AuthorizationProcessor processor = authProcessor;
+    if (processor == null) {
+      return super.revoke(performer, permissions, resource, revokee, grantModes);
+    }
+
+    Collection<CompletableFuture<Void>> stages = new ArrayList<>(grantModes.length);
+
+    for (GrantMode grantMode : grantModes) {
+      Pair<PermissionKind, AuthorizationOutcome> mode = grantMode(grantMode);
+      CompletionStage<Void> stage =
+          processor.addPermissions(
+              ImmutableActor.of(performer.getName()),
+              mode.getValue1(),
+              mode.getValue0(),
+              permissions(permissions),
+              resource(resource),
+              role(revokee));
+
+      stages.add(stage.toCompletableFuture());
+    }
+
+    get(CompletableFuture.allOf(stages.toArray(new CompletableFuture[0])));
+
+    // assume all permissions were revoked if we did not an exception
+    return permissions;
+  }
+
+  private static void get(CompletionStage<Void> stage) {
+    try {
+      // wait for completion since the calling API is synchronous
+      stage.toCompletableFuture().get();
+    } catch (ExecutionException e) {
+      Throwable cause = e.getCause();
+      if (cause instanceof RequestValidationException) {
+        throw (RequestValidationException) cause;
+      }
+
+      throw new AuthRequestExecutionException(SERVER_ERROR, e.getMessage(), e);
+    } catch (Exception e) {
+      throw new AuthRequestExecutionException(SERVER_ERROR, e.getMessage(), e);
+    }
+  }
+
+  private static Pair<PermissionKind, AuthorizationOutcome> grantMode(GrantMode grantMode) {
+    switch (grantMode) {
+      case GRANTABLE:
+        return Pair.with(PermissionKind.AUTHORITY, AuthorizationOutcome.ALLOW);
+
+      case GRANT:
+        return Pair.with(PermissionKind.ACCESS, AuthorizationOutcome.ALLOW);
+
+      case RESTRICT:
+        return Pair.with(PermissionKind.ACCESS, AuthorizationOutcome.DENY);
+
+      default:
+        throw new UnsupportedOperationException("Unsupported grant mode: " + grantMode);
+    }
+  }
+
+  private static EntitySelector role(RoleResource roleResource) {
+    if (!roleResource.hasParent()) { // root == all roles
+      return EntitySelector.wildcard();
+    } else {
+      return EntitySelector.byName(roleResource.getRoleName());
+    }
+  }
+
+  private static AuthorizedResource resource(IResource resource) {
+    String resourcePath = resource.getName();
+    ImmutableAuthorizedResource authResource;
+
+    if (resource instanceof FunctionResource) {
+      authResource = ImmutableAuthorizedResource.of(ResourceKind.FUNCTION);
+
+      FunctionResource fr = (FunctionResource) resource;
+
+      boolean hasKeyspace = resourcePath.indexOf("/") > 0;
```

Comment:
I don't understand why `resourcePath.indexOf("/") > 0` means that it has a keyspace. Could you elaborate more?

Reply:
It's based on the way `getName` output is formatted. Unfortunately, `FunctionResource` does not provide a getter to checking whether it has a keyspace and will throw exceptions if we try to get the keyspace if it does not have one.